File: Matrix Manipulation/51. N-Queens.cpp
class Solution {
  public:
      static bool validpuzzle(vector<vector<int>>& arr, int i, int j){
          for (int k=0;k<arr.size();k++){
              if(arr[k][1]==j){return false;}
              else if (arr[k][1]-j==arr[k][0]-i){return false;}
              else if (arr[k][1]-j==i-arr[k][0]){return false;}
          }
          return true;
      }
      vector<string> init(int n){
          vector<string> arr ;
          string tmp;
          for (int i=0;i<n;i++){
              tmp = "";
              for (int j=0;j<n;j++){
                  tmp +=".";
              }
              arr.push_back(tmp);
          }
          return arr;
      }
      void completepuzzle(vector<vector<int>>& arr, int i, int n,vector<vector<string>>& alls){
          if(i==n){
              vector<string> new_elt = init(n);
              int pos_x,pos_y;
              for (int k = 0; k < arr.size(); k++) {
                  pos_x = arr[k][0];
                  pos_y = arr[k][1];
                  new_elt[pos_x][pos_y] = 'Q';
              }
              alls.push_back(new_elt);
          }
          for (int j=0;j<n;j++){
              if (validpuzzle(arr,i,j)){
                  vector<vector<int>> new_arr = arr;
                  new_arr.push_back({i,j});
                  completepuzzle(new_arr,i+1,n,alls);
              }
          }
      }
      vector<vector<string>> solveNQueens(int n) {
          vector<vector<string>> alls;
          vector<vector<int>> arr;
          completepuzzle(arr,0,n,alls);
          return alls;
     
      }
};
// Time complexity O(n^n) ans spade complexity(n!)

File: Matrix Manipulation/51. N-Queens.py
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """

        def validpuzzle(arr,i,j):
            for x,y in arr: 
                if y==j:
                    return False
                elif y-j==x-i:
                    return False
                elif y-j==i-x:
                    return False

            return True
        
        def complete(arr,i,alls):
            if i==n :
                new = ["."*n for _ in range(n)]
                for x,y in arr:
                    new[x] = new[x][:y]+"Q"+new[x][y+1:]
                alls.append(new)               
            
            else :
                for j in range(n):
                    if validpuzzle(arr,i,j):
                        complete(arr+[(i,j)],i+1,alls)
        
        alls = []
        complete([],0,alls)
        return alls
      
# Time complexity O(n^n) and space complexity O(n!)
